Wraps transaction into separate class

mitchhentges opened the pull request:
This breaks apart `EditService`:
* Transactions (including committing) are now represented by:
`with WritableGooglePlay.transaction() as google_play:`
* Managing real vs mock API has been split into the `GooglePlayConnection` abstraction
* Read-only logic is simpler than write logic (no need to commit after), so it's split into a different class and made easier to use.

-----

@JohanLorenzo I'm pretty happy with these changes, though I'm on the fence about splitting `ReadOnlyGooglePlay` from `WritableGooglePlay`. My main reasoning for doing happened when I saw this:
```
with GooglePlay.transaction(connection, 'org.mozilla.firefox', do_not_commit=True) as google_play:
    check_rollout(google_play, config.days)
```
When you're doing work that just requires read-only access, it's noisy to add `do_not_commit=True`, and setting up the whole `with` statement is annoying. I'm especially worried that it will be too easy to omit the `do_not_commit` parameter, causing read-only work to take longer, since an unnecessary commit happens.
So, by splitting the read-only class from the writable one, we get:
```
google_play = ReadOnlyGooglePlay.create(connection, 'org.mozilla.firefox')
check_rollout(google_play, config.days)
```
Devs will realize the existence of `ReadOnlyGooglePlay` (if there's a `WriteableGooglePlay`, there's probably a "read"-related one), and will opt to use it if possible, since its less verbose to write. I think this is way better than having to scan for missing `do_not_commit` parameters or doing unnecessary commits. What do you think?

JohanLorenzo commented on mozapkpublisher/push_apk.py:
```diff
@@ -62,43 +63,22 @@ def push_apk(
-    for path, metadata in apks_metadata_per_paths.items():
-        edit_service.upload_apk(path)
+        with WritableGooglePlay.transaction(connection, package_name,
+                                            do_not_commit=not commit) as google_play:
```
Nit: Let's avoid negative boolean names. It adds a mental layer of indirection when reading the code.

mitchhentges commented:
Hey, thanks for the thoughtful response :)

A couple thoughts:
1. IMHO, I'd rather make two calls than pass unnecessary parameters to a function. For example:
```
with WritableGooglePlay(MockGooglePlayConnection(), package_name) as google_play:
// vs
with WritableGooglePlay(package_name, None, None, contact_google_play=False) as google_play:
```
This is for a couple reasons:
* Having mutually exclusive parameters adds more noise to the calling interface. It forces readers to consider whether the `None` values are meant to represent anything, or if they're just filler values
* It's more explicit to provide that separate function call. Theoretically, the `contact_google_play` flag represents the same information as the second function name (`MockGooglePlayConnection()`). And, since the relevant parameters are grouped up front, it's more clear what's used for what, without having to grub through docs.

```
WritableGooglePlay(package_name, GooglePlayConnection('me@gmail.com', 'file.p12'))
                                                                  |     |
                    these two parameters are related, and it's    |    /
                    clear because they're called in a function    -----    
                    together

WritableGooglePlay(package_name, 'me@gmail.com', 'file.p12', True)
                                             |          |    |
                                             |          /    /
            These parameters are all related ----------------
            but it's not obvious without consulting docs
```

2. I'm not sure that the `mode='w/r'` style makes sense for this API. The biggest up-front cost of this decision would be that it isn't obvious what kind of object is returned by the single `GooglePlay` function without looking at the docs/source code. In the case of `ZipFile`, the `mode` style may have made sense since it's lowkey mirroring a file system api (which takes that parameter), as well as since there's a single `ZipFile` class (unlike the multiple for Google Play).
Additionally, since there's a one-to-one mapping between `mode` and class name, I'm unsure what value this is providing
```
mode='r' # => ReadOnlyGooglePlay(...)
mode='w' # => WritableGooglePlay(...)
```
One other note is that, if at some point in the future, the two implementations require a different set of parameters (perhaps `WritableGooglePlay` would need an extra credential, since it performs more powerful operations? Bad example, but w/e :smile:), then it would make the single function more leaky, since there'd be an optional parameter that's only used depending on the value in `mode`.

3. That's enough concerns outta me though, I _really_ like your point that the static `.transaction(...)`/`.create(...)` functions not being pythonic, that's a good observation.
I'll make `WritableGooglePlay` a context manager class, so you can just:
```
with WritableGooglePlay(...) as google_play:
```
and I'll add a `commit(...)` function.

This way, it's more like `ZipFile` and others where you have the option of using it as a context manager, but if you want to manually open and close it, you can:
```
google_play = WritableGooglePlay(...) 
# ...
google_play.commit()
```

mitchhentges commented:
:thinking: Looking at this a little further, I'm not sure it makes sense to remove the static function for each of the `GooglePlay` implementations. Let's focus on `ReadOnlyGooglePlay` here:
Since there's no clean up possible for the read-only connection, it doesn't make sense to use with a context manager.

However, there _is_ a set-up sequence, where a connection is used to establish a transaction. Importantly, this operation is a "side-effect"-y operation, so it doesn't make sense to do in the constructor.
The cleanest of of managing this is by having a function that takes required parameters, performs the side-effect-y operation, then creates the object and returns it:
```
def create(connection, package_name):
    edit_resource = connection.spool_it_up(package_name)
    edit_id = edit_resource.activate_turbo()
    return ReadOnlyGooglePlay(edit_resource, edit_id, package_name)
```

And, once we have this function that is closely related to `ReadOnlyGooglePlay`, we either name it to show that they're related (`create_read_only_google_play(conn, package_name)`) or we namespace it (`ReadOnlyGooglePlay.create(conn, package_name)`).

:thinking: That's the costs and benefits affecting my design decisions so far, what do you think?

JohanLorenzo commented on mozapkpublisher/common/googleplay.py:
```diff
@@ -159,48 +204,16 @@ def update_whats_new(self, language, apk_version_code, whats_new):
+            logger.info('Changes committed')
+            logger.debug('edit_id "{}" for "{}" has been committed'.format(edit_id, package_name))
+        else:
+            logger.warning('Transaction not committed, since `commit` was `False`')
```
Followup: Use `validate()` when we don't commit https://github.com/googleapis/google-api-python-client/blob/715bd7f9ca9a324a05ee05a6476af383e54930f9/docs/dyn/androidpublisher_v3.edits.html#L135 (I learned about this function last week)

mitchhentges replied:
I'd like to make this an independent change, since it's less related to this refactor and I don't want to increase the surface area of issues

JohanLorenzo commented on mozapkpublisher/test/test_push_apk.py:
```diff
@@ -27,11 +29,11 @@
 @pytest.fixture
-def edit_service_mock():
-    return create_autospec(googleplay.EditService)
+def writable_google_play_mock():
```
Nit: There's no `writable` anymore

mitchhentges commented on mozapkpublisher/common/googleplay.py:
```diff
@@ -41,57 +43,102 @@ def add_general_google_play_arguments(parser):
+    """Represents an "edit" to an app on the Google Play store
 
-        self._edit_id = None
+    Create an instance by calling GooglePlayEdit.transaction(), instead of using the
```
I don't understand: you're advocating that `transaction()` be moved into a separate function?
Since it's so closely related to `GooglePlayEdit`, what are we gaining by separating it?

JohanLorenzo replied:
To me, if you don't want future developers to call the constructor of `GooglePlayEdit` directly, there is a pythonic way:

You can expose the function in the module itself. This way, developers do:

```python
from mozapkpublisher.common import googleplay

with googleplay.transaction() as google_play:
    pass
```

This what happens in several python modules. They usually don't name space the function down to the class. For instance: 
   - https://docs.python.org/3/library/warnings.html#testing-warnings
   - https://docs.python.org/3/library/shelve.html#shelve.open 
   - https://docs.python.org/3/library/urllib.request.html#examples

This way, `transaction()` appears at the top of the documentation, so you don't document the class first, so you don't need to tell developers to not call the constructor directly.

mitchhentges commented on mozapkpublisher/push_apk.py:
```diff
@@ -62,43 +63,21 @@ def push_apk(
     )
 
+    # TODO make programmatic usage of this library provide a "GooglePlayConnection" object, rather
+    # than having to provide redundant information like "service_account" and "credentials" when
```
This TODO was created before that RFC discussion, yeah

mitchhentges commented on mozapkpublisher/common/googleplay.py:
```diff
@@ -159,48 +204,16 @@ def update_whats_new(self, language, apk_version_code, whats_new):
-    return service
+    @staticmethod
+    @contextmanager
+    def transaction(connection, package_name, commit):
```
Should we leave it explicit? I definitely don't believe that we should default it to `true`, but I think that requiring devs to specify a value is a good thing. It reduces confusion when devs don't realize that the parameter is possible ("why aren't my changes saving?), but requires an explicit `true` if devs are confident that they want to apply their changes

JohanLorenzo replied:
In addition to avoiding footguns, defaulting this param has another pro to me. `commit` becomes a `kwarg`, so if it's an incentive to call it this way:
```python
with googleplay.transaction(connection, 'some.package.name', commit=True) as gp:
    pass
```

rather than that way:

```python
with googleplay.transaction(connection, 'some.package.name', True) as gp:
    # What does True mean? 
    pass
```

While, I agree it doesn't prevent the latter from happening. It's just an incentive. 

To me, the latter is more dangerous and can lead newcomers to just copy and paste an existing (and working) line while having very few context about what this `True` is about. I think the damage done by a committed transaction outweighs the damage done by the "why aren't my changes being applied" scenario. Moreover, the reason why changes aren't applied is explicitly called out in the logs. So guessing the fix shouldn't be a hard task.

mitchhentges replied:
I think I've found the best of both words: I've forced it to be a kwarg without giving it a default value:

```
def edit(connection, package_name, *, commit):
    ...
```

JohanLorenzo replied:
Oooooooh! TIL https://www.python.org/dev/peps/pep-3102/

That's an excellent finding 😃

mitchhentges commented on mozapkpublisher/common/googleplay.py:
```diff
@@ -41,57 +43,102 @@ def add_general_google_play_arguments(parser):
+        return edit_service_mock
+
+
+def connection_for_options(contact_google_play, service_account, credentials_file):
```
I'm not sure that we'd get any value out of defaulting `contact_google_play` to `True` - it's only used within `mozapkpublisher`, and each of those usages don't have a constant value for `contact_google_play`. This would make the "defaulting" redundant, so I'm not sure it would provide any value

JohanLorenzo replied:
True. Okay to keep it as is.

mitchhentges commented on mozapkpublisher/push_apk.py:
```diff
@@ -62,43 +63,21 @@ def push_apk(
-
-    for path, metadata in apks_metadata_per_paths.items():
-        edit_service.upload_apk(path)
+        with GooglePlayEdit.transaction(connection, package_name, commit) as google_play:
```
Good point, if `connection` isn't provided at a higher level, then this absolutely makes sense.
Let's wait for [this RFC discussion](https://github.com/mozilla-releng/releng-rfcs/pull/29) to clear up, then I'll update this accordingly

mitchhentges replied:
Updated to send raw parameters

mitchhentges commented on mozapkpublisher/common/googleplay.py:
```diff
@@ -41,57 +43,102 @@ def add_general_google_play_arguments(parser):
+
+class MockGooglePlayConnection:
+    @staticmethod
+    def get_edit_resource():
```
Interesting - why should this be idempotent? It's not idempotent for the "real" `GooglePlayConnection`

MihaiTabara commented:
Dropping some initial thoughts on this. 

To be honest, I like both options, for different reasons. But overall, I'm slightly leaning towards the pythonic way of doing it, rather than classes.

Reasons I like the using classes approach:
1. solves the extra parameters noise which can be quite annoying to follow in the other case of overloading the function with multiple functionalities
2. no-one likes to toggle between using an object and the docs, ideally things should be self-documenting; objects are easier to read, reducing the time to go back and forth and double-check if a certain piece of logic from the docs validates
4. fits better in the future space of a potential, more mainstream, Python with static-typing hinting (3.7+ onwards)

Reasons I like the pythonic way:
1. this is the pythonic way; most of the widely used modules, like `requests` or `flask` are written in that manner
2. in a more generic perspective; altough types are getting more and more traction, we're not yet using them in a wide perspective in our codebase. This is, by no means, an "but this is how we've always done it" scenario, but there may be some underlying reasoning for that. I guess while the hint typing is something that was already introduced, Python will always remain a dynamically typed language so I believe we should stick to the primitives as much as we can. Since static typing is still under development in Python, it is subject to change in its future releases.
3. slightly easier to test/mock. classes add import overhead

With all that said, I still believe consistency is good across an organization so I'm leaning towards the pythonic way of overloading multiple behaviors within the primitive functions.

mitchhentges commented on mozapkpublisher/common/googleplay.py:
```diff
@@ -41,57 +43,102 @@ def add_general_google_play_arguments(parser):
-        else:
-            logger.warning('`commit` option was not given. Transaction not committed.')
+    def get_edit_resource(self):
+        return self._edit_resource
```
Good call!

mitchhentges commented:
Thanks Mihai for weighing in on the discussion. Though I'm not super hyped on the result, I'm glad that we can unblock this PR and move forward again :smile:

-----

> this is the pythonic way; most of the widely used modules, like requests or flask are written in that manner

Classes aren't less pythonic :thinking: 
You mentioned two examples of dictionary use within the python ecosystem: `requests` and `flask`. In those specific cases, they use dictionaries for raw HTTP data because it _has_ to be dynamic (data at the raw HTTP layer has to be unstructured by definition :smile:)
[Here in my RFC](https://github.com/mozilla-releng/releng-rfcs/pull/29/files#diff-d70f1afa8b34b70a22042da538a842c7R95), I mean to address this specific case, and how I agree that dynamic data should be held in a dynamic container.

However, in an API boundary with a known set of possible data, it's now **possible** to represent your data in an explicit structure.

> Python will always remain a dynamically typed language so I believe we should stick to the primitives as much as we can.

Yes, it's dynamically typed, but Python is giving us the _built-in_, _native_ ability to structure known types of data. An argument analogous to this is "computers will always remain operating on primitive processor instructions, so we should program at that primitive level accordingly."

> Since static typing is still under development in Python, it is subject to change in its future releases.

Using classes to encapsulate logic and config has been possible since before the release of the first Python 3. They aren't experimental at this point.

> classes add import overhead

The amount of weight that this point appears to have has surprised me.

-----

Mihai and Johan, you guys are pretty aligned on the direction here, so I'll :man_shrugging:. I've updated the PR to use raw primitives and removed the `GooglePlayConnection` abstraction, and removed the two follow-up "abstraction" PRs.

I'm not confident in the direction that we'll take, but I'm really happy to be able to have a clear direction for our team that I can implement this PR against so we can have it landed :smiley:

JohanLorenzo commented on CHANGELOG.md:
```diff
@@ -2,6 +2,11 @@
+## [upcoming]
+
+### Changed
+* Uses RAII to represent operations that require an "edit"
```
TIL this acronym 👍

JohanLorenzo commented on mozapkpublisher/check_rollout.py:
```diff
@@ -40,14 +41,11 @@ def main():
-    for (release, age) in check_rollout(edit_service, config.days):
-        print('fennec {} is on staged rollout at {}% but it shipped {} days ago'.format(
-              release['name'], int(release['userFraction'] * 100), int(age / DAY)))
+    with googleplay.edit(True, config.service_account, config.google_play_credentials_file.name,
```
Nit: Let's make `contact_google_play` a mandatory kwarg too. `True` is ambiguous here.

JohanLorenzo commented on mozapkpublisher/common/googleplay.py:
```diff
@@ -160,47 +145,57 @@ def update_whats_new(self, language, apk_version_code, whats_new):
+        logger.info('Changes committed')
+        logger.debug('edit_id "{}" for "{}" has been committed'.format(edit_id, package_name))
+    else:
+        edit_resource.validate()
```
\o/ 👍

mitchhentges replied:
Err, I didn't want to add this yet, since I want to test this new interaction with Google Play separately. I'll roll it back :)

JohanLorenzo commented on mozapkpublisher/push_apk.py:
```diff
@@ -65,40 +65,14 @@ def push_apk(
-
-    for path, metadata in apks_metadata_per_paths.items():
-        edit_service.upload_apk(path)
+        with googleplay.edit(contact_google_play, service_account, google_play_credentials_file.name,
```
I like the name of the function! 🙂
